example/babelstream/src/AlpakaStream.cpp

```diff
@@ -0,0 +1,258 @@
+{
+    auto const workdiv = WorkDiv{arraySize / TBSIZE, TBSIZE, 1};
+    // auto const workdiv = alpaka::getValidWorkDiv(devAcc, arraySize);
+    alpaka::exec<Acc>(
+        queue,
+        workdiv,
+        TriadKernel{},
+        alpaka::getPtrNative(d_a),
+        alpaka::getPtrNative(d_b),
+        alpaka::getPtrNative(d_c));
+    alpaka::wait(queue);
+}
+
+struct NstreamKernel
+{
+    template<typename TAcc, typename T>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, T* a, T const* b, T const* c) const
+    {
+        const T scalar = startScalar;
+        auto const [i] = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        a[i] += b[i] + scalar * c[i];
+    }
+};
+
+template<typename T>
+void AlpakaStream<T>::nstream()
+{
+    auto const workdiv = WorkDiv{arraySize / TBSIZE, TBSIZE, 1};
+    // auto const workdiv = alpaka::getValidWorkDiv(devAcc, arraySize);
+    alpaka::exec<Acc>(
+        queue,
+        workdiv,
+        NstreamKernel{},
+        alpaka::getPtrNative(d_a),
+        alpaka::getPtrNative(d_b),
+        alpaka::getPtrNative(d_c));
+    alpaka::wait(queue);
+}
+
+struct DotKernel
+{
+    template<typename TAcc, typename T>
+    ALPAKA_FN_ACC void operator()(TAcc const& acc, T const* a, T const* b, T* sum, int arraySize) const
+    {
+        // TODO - test if sharedMem bug is affecting performance here
+        auto& tb_sum = alpaka::declareSharedVar<T[TBSIZE], __COUNTER__>(acc);
+
+        auto [i] = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
+        auto const [local_i] = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc);
+        auto const [totalThreads] = alpaka::getWorkDiv<alpaka::Grid, alpaka::Threads>(acc);
+
+        tb_sum[local_i] = 0.0;
+        for(; i < arraySize; i += totalThreads)
+            tb_sum[local_i] += a[i] * b[i];
+
+        auto const [blockDim] = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc);
+        for(int offset = blockDim / 2; offset > 0; offset /= 2)
+        {
+            alpaka::syncBlockThreads(acc);
+            if(local_i < offset)
+                tb_sum[local_i] += tb_sum[local_i + offset];
+        }
+
+        auto const [blockIdx] = alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc);
+        if(local_i == 0)
+            sum[blockIdx] = tb_sum[local_i];
+    }
+};
+
+template<typename T>
+T AlpakaStream<T>::dot()
+{
+    auto const workdiv = WorkDiv{DOT_NUM_BLOCKS, TBSIZE, 1};
+    // auto const workdiv = alpaka::getValidWorkDiv(devAcc, DOT_NUM_BLOCKS * TBSIZE);
+    alpaka::exec<Acc>(
+        queue,
+        workdiv,
+        DotKernel{},
+        alpaka::getPtrNative(d_a),
+        alpaka::getPtrNative(d_b),
+        alpaka::getPtrNative(d_sum),
+        arraySize);
+    alpaka::wait(queue);
+
+    alpaka::memcpy(queue, sums, d_sum);
+    T const* sumPtr = alpaka::getPtrNative(sums);
+    // TODO(bgruber): replace by std::reduce, when gcc 9.3 is the baseline
+    return std::accumulate(sumPtr, sumPtr + DOT_NUM_BLOCKS, 0);
```

Comment:
Needs to be: `return std::accumulate(sumPtr, sumPtr + DOT_NUM_BLOCKS, static_cast<T>(0));` otherwise it cast each value to `int`.

Reply:
Thx! Fixed.